### ISSUE TITLE
Support for IR Air Conditioner

### ISSUE THREAD

**Issue by donavanbecker** [OpenWonderLabs/homebridge-switchbot-openapi#55]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

```
[SwitchBot] {"deviceId":"02-202101151706-53910675","deviceName":"Projector","remoteType":"Projector","hubDeviceId":"D949E636A458"}
16/01/2021, 19:19:49 [SwitchBot] Device: Projector with Device Type: Projector, is currently not supported. Submit Feature Requests Here: https://git.io/JL14Z
16/01/2021, 19:19:49 [SwitchBot] {"deviceId":"02-202101161909-84727995","deviceName":"Air Conditioner MB","remoteType":"Air Conditioner","hubDeviceId":"D949E636A458"}
16/01/2021, 19:19:49 [SwitchBot] Device: Air Conditioner MB with Device Type: Air Conditioner, is currently not supported. Submit Feature Requests Here: https://git.io/JL14Zk```

**Comment by donavanbecker** [comment]
@willliamchan,

I have opened this to keep track of the air conditioner request.

**Comment by willliamchan** [comment]
Sure thing

Regards
William
________________________________

**Comment by joshepw** [comment]
Please add Air Conditioner Support, Maybe with the Heater Cooler Service you can change Fan Speed, Fan Swing, PowerOn/Off, and config the min/max device like: 
[Characteristic].setProps({
            minValue: 16, <-- Config param
            maxValue: 30, <-- Config param
            minStep: 1
})

**Comment by joshepw** [comment]
Hello again dude, i made this AirConditioner.ts for irDevices, i forked your bera Repo, can try add this or modify:

`import {
	Characteristic,
	CharacteristicEventTypes,
	CharacteristicGetCallback,
	CharacteristicValue,
	PlatformAccessory,
	Service,
} from 'homebridge';
import { SwitchBotPlatform } from '../platform';
import { DeviceURL } from '../settings';
import { irdevice, deviceStatusResponse } from '../configTypes';

/**
 * Platform Accessory
 * An instance of this class is created for each accessory your platform registers
 * Each accessory may expose multiple services of different service types.
 */
export class AirConditioner {
	static readonly MODE_AUTO = 1;
	static readonly MODE_COOL = 2;
	static readonly MODE_DRY = 3;
	static readonly MODE_FAN = 4;
	static readonly MODE_HEAT = 5;

	service!: Service;

	Active!: CharacteristicValue;
	ActiveIdentifier!: CharacteristicValue;
	deviceStatus!: deviceStatusResponse;

	RotationSpeed: any;
	lastTemperature: number | undefined;
	currentTemperature: number | undefined;
	currentMode: number | undefined;
	currentFanSpeed: number | undefined;
	busy!: boolean;

	timeout:any = null;

	constructor(
		private readonly platform: SwitchBotPlatform,
		private accessory: PlatformAccessory,
		public device: irdevice,
	) {
		// set accessory information
		this.accessory
			.getService(this.platform.Service.AccessoryInformation)!
			.setCharacteristic(this.platform.Characteristic.Manufacturer, 'SwitchBot')
			.setCharacteristic(this.platform.Characteristic.Model, this.device.remoteType)
			.setCharacteristic(this.platform.Characteristic.SerialNumber, this.device.deviceId);

		// get the Television service if it exists, otherwise create a new Television service
		// you can create multiple services for each accessory
		(this.service =
			this.accessory.getService(this.platform.Service.HeaterCooler) || this.accessory.addService(this.platform.Service.HeaterCooler)),
			`${this.device.deviceName} ${this.device.remoteType}`;

		// To avoid "Cannot add a Service with the same UUID another Service without also defining a unique 'subtype' property." error,
		// when creating multiple services of the same type, you need to use the following syntax to specify a name and subtype id:
		// this.accessory.getService('NAME') ?? this.accessory.addService(this.platform.Service.Outlet, 'NAME', 'USER_DEFINED_SUBTYPE');

		// set the service name, this is what is displayed as the default name on the Home app
		// in this example we are using the name we stored in the `accessory.context` in the `discoverDevices` method.
		this.service.setCharacteristic(
			this.platform.Characteristic.Name,
			`${this.device.deviceName} ${this.device.remoteType}`,
		);

		// handle on / off events using the Active characteristic
		this.service
			.getCharacteristic(this.platform.Characteristic.Active)
			.on(CharacteristicEventTypes.SET, (value: any, callback: CharacteristicGetCallback) => {
				this.platform.log.debug('AirConditioner %s Set Active: %s', this.accessory.displayName, value);
				this.platform.log.warn(value);

				try {
					if (value === this.platform.Characteristic.Active.INACTIVE) {
						this.pushAirConditionerOffChanges();
					} else {
						this.pushAirConditionerOnChanges();
					}
	
					this.Active = value;
					this.service.updateCharacteristic(this.platform.Characteristic.Active, this.Active);
					callback(null);
				} catch (error) {
					callback(error);
				}
			});

		this.service.getCharacteristic(this.platform.Characteristic.CurrentTemperature)
			.setProps({
				minValue: 0,
				maxValue: 100,
				minStep: 0.01
			})
			.on('get', this.handleCurrentTemperatureGet.bind(this));

		this.service.getCharacteristic(this.platform.Characteristic.TargetHeaterCoolerState)
			.on('set', this.setMode.bind(this));

		this.service.getCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState)
			.on('get', this.getCurrentHeaterCoolerState.bind(this));

		this.service.getCharacteristic(this.platform.Characteristic.HeatingThresholdTemperature)
			.setProps({
				minValue: 16,
				maxValue: 30,
				minStep: 1
			})
			.on('get', this.getHeatingUpOrDwTemperature.bind(this))
			.on('set', this.setHeatingUpOrDwTemperature.bind(this));

		this.service.getCharacteristic(this.platform.Characteristic.CoolingThresholdTemperature)
			.setProps({
				minValue: 16,
				maxValue: 30,
				minStep: 1
			})
			.on('get', this.getHeatingUpOrDwTemperature.bind(this))
			.on('set', this.setHeatingUpOrDwTemperature.bind(this));
		
		this.service
			.getCharacteristic(this.platform.Characteristic.RotationSpeed)
			.setProps({
				minStep: 1,
				minValue: 1,
				maxValue: 4,
			})
			.on('get', (callback) => {
				if (!this.currentFanSpeed || this.currentFanSpeed === 1) {
					callback(null, 4);
				} else {
					callback(null, this.currentFanSpeed - 1);
				}
			})
			.on('set', (value: any, callback) => {
				this.platform.log.debug('Air Fan %s Set Active: %s', this.accessory.displayName, value);
				this.platform.log.warn(value);

				if (value === 4) {
					this.currentFanSpeed = 1;
				} else {
					this.currentFanSpeed = value + 1;
				}
				
				this.pushAirConditionerStatusChanges();

				this.service.updateCharacteristic(this.platform.Characteristic.RotationSpeed, this.currentFanSpeed || 1);

				callback(null, this.currentFanSpeed || 1);
			});
	}

	getHeatingUpOrDwTemperature(callback) {
		callback(null, this.currentTemperature || 24);
	}
		
	setHeatingUpOrDwTemperature(temp, callback) {
		try {
			this.pushAirConditionerStatusChanges();
			this.lastTemperature = this.currentTemperature;
			this.currentTemperature = parseInt(temp);
		} catch (error) {
			this.platform.log.error(error);
			callback(error);
		}
	}

	getCurrentHeaterCoolerState(callback) {
		if (this.Active === 1) {
			if ((this.currentTemperature || 24) < (this.lastTemperature || 30)) {
				callback(null, this.platform.Characteristic.CurrentHeaterCoolerState.COOLING);
			} else {
				callback(null, this.platform.Characteristic.CurrentHeaterCoolerState.HEATING);
			}
		} else {
			callback(null, this.platform.Characteristic.CurrentHeaterCoolerState.INACTIVE);
		}
	}

	setMode(state, callback) {
		switch (state) {
			case this.platform.Characteristic.TargetHeaterCoolerState.AUTO:
				this.currentMode = AirConditioner.MODE_AUTO;
				break;

			case this.platform.Characteristic.TargetHeaterCoolerState.COOL:
				this.currentMode = AirConditioner.MODE_COOL;
				break;

			case this.platform.Characteristic.TargetHeaterCoolerState.HEAT:
				this.currentMode = AirConditioner.MODE_HEAT;
				break;
		
			default:
				break;
		}

		this.pushAirConditionerStatusChanges();

		callback(null);
	}

	/**
	 * Handle requests to get the current value of the "Current Temperature" characteristic
	 */
	handleCurrentTemperatureGet(callback) {
		this.platform.log.debug('Triggered GET CurrentTemperature');
		
		this.service.getCharacteristic(this.platform.Characteristic.CurrentTemperature).updateValue(this.currentTemperature || 24);

		callback(null, this.currentTemperature || 24);
	}

	/**
	 * Pushes the requested changes to the SwitchBot API
	 * deviceType				commandType     Command	          command parameter	         Description
	 * AirConditioner:        "command"       "swing"          "default"	        =        swing
	 * AirConditioner:        "command"       "timer"          "default"	        =        timer
	 * AirConditioner:        "command"       "lowSpeed"       "default"	        =        fan speed to low
	 * AirConditioner:        "command"       "middleSpeed"    "default"	        =        fan speed to medium
	 * AirConditioner:        "command"       "highSpeed"      "default"	        =        fan speed to high
	 */
	async pushAirConditionerOnChanges() {
		if (this.Active !== 1) {
			const payload = {
				commandType: 'command',
				parameter: 'default',
				command: 'turnOn',
			} as any;

			await this.pushChanges(payload);
		}
	}

	async pushAirConditionerOffChanges() {
		const payload = {
			commandType: 'command',
			parameter: 'default',
			command: 'turnOff',
		} as any;

		await this.pushChanges(payload);
	}

	pushAirConditionerStatusChanges() {
		if (!this.busy) {
			this.busy = true;
			this.service.updateCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState, this.platform.Characteristic.CurrentHeaterCoolerState.IDLE);
		}

		clearTimeout(this.timeout);

		// Make a new timeout set to go off in 1000ms (1 second)
		this.timeout = setTimeout(this.pushAirConditionerDetailsChanges.bind(this), 1500);
	}

	pushAirConditionerDetailsChanges() {
		const payload = {
			commandType: 'command',
			parameter: `${this.currentTemperature || 24},${this.currentMode || 1},${this.currentFanSpeed || 1},${this.Active === 1 ? 'on':'off'}`,
			command: 'setAll',
		} as any;
		
		if (this.Active === 1) {
			if ((this.currentTemperature || 24) < (this.lastTemperature || 30)) {
				this.service.updateCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState, this.platform.Characteristic.CurrentHeaterCoolerState.COOLING);
			} else {
				this.service.updateCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState, this.platform.Characteristic.CurrentHeaterCoolerState.HEATING);
			}
		} else {
			this.service.updateCharacteristic(this.platform.Characteristic.CurrentHeaterCoolerState, this.platform.Characteristic.CurrentHeaterCoolerState.INACTIVE);
		}

		this.pushChanges(payload);
	}

	public async pushChanges(payload: any) {
		this.platform.log.info(
			'Sending request for',
			this.accessory.displayName,
			'to SwitchBot API. command:',
			payload.command,
			'parameter:',
			payload.parameter,
			'commandType:',
			payload.commandType,
		);
		this.platform.log.debug('Air Conditioner %s pushChanges -', this.accessory.displayName, JSON.stringify(payload));

		// Make the API request
		const push = await this.platform.axios.post(`${DeviceURL}/${this.device.deviceId}/commands`, payload);
		this.platform.log.debug('Air Conditioner %s Changes pushed -', this.accessory.displayName, push.data);

		this.busy = false;
	}
}
`

**Comment by donavanbecker** [comment]
@joshepw Thanks! If you would like to submit a PR towards the beta branch, I can merge this? Or I can put it into the already started code for airconditioners. Whatever you decide. I am working when I have time to get all device implemented. I am getting close.

**Comment by donavanbecker** [comment]
> @joshepw Thanks! If you would like to submit a PR towards the beta branch, I can merge this? Or I can put it into the already started code for airconditioners. Whatever you decide. I am working when I have time to get all device implemented. I am getting close.

I took parts of what you posted, and will issue a new beta soon.

**Comment by donavanbecker** [comment]
@willliamchan, @joshepw and @amanolo I have added support for Air Conditioners thanks for helping @joshepw, made this quicker.

if you can test the latest beta?
Beta: `v2.0.1-beta.7`

**Comment by joshepw** [comment]
Hi @donavanbecker , can you write me to telegram? My user is @joshepw and i can help you with refactor some code.

Can you explain me how test the beta in my homebridge? I wrote the code without test.

**Comment by donavanbecker** [comment]
As you probably already know, this has been released with v2.1.0
